### ISSUE TITLE
e2e flake: Kubelet syncloop hangs

### ISSUE THREAD

**Issue by wojtek-t** [kubernetes/kubernetes#22259]
In one of the the kubemark runs, one node of the base cluster (real Kubelet) became unready. However, this was preceded by a bunch of following logs:

    I0229 18:36:47.747404    3429 manager.go:451] Pod status is inconsistent with cached status for pod "fluentd-elasticsearch-jenkins-e2e-minion-a6cx_kube-system(b409980b-df12-11e5-9e22-42010af00002)", a reconciliation should be triggered:
     {"phase":"Pending"

    A: }

    B: ,"conditions":[{"type":"Ready","status":"False","lastProbeTime":null,"lastTransitionTime":"2016-02-29T18:31:52Z","reason":"ContainersNotReady","message":"containers with unready status: [fluentd-elasticsearch]"}],"hostIP":"10.240.0.4","startTime":"2016-02-29T18:31:52Z","containerStatuses":[{"name":"fluentd-elasticsearch","state":{"waiting":{"reason":"ContainerCreating","message":"Image: gcr.io/google_containers/fluentd-elasticsearch:1.14 is ready, container is creating"}},"lastState":{},"ready":false,"restartCount":0,"image":"gcr.io/google_containers/fluentd-elasticsearch:1.14","imageID":""}]}

My feeling is that it may be related to the following logs:

    W0229 18:36:59.248357    4318 kubelet.go:2741] ConfigureCBR0 requested, but PodCIDR not set. Will not configure CBR0 right now
    ...
    I0229 18:36:59.450359    4318 kubelet.go:1131] Node jenkins-e2e-minion-a6cx was previously registered
    I0229 18:36:59.452361    4318 kubelet.go:3557] Setting Pod CIDR:  -> 10.245.2.0/24
    I0229 18:36:59.524828    4318 kubelet.go:2715] Recording NodeNotReady event message for node jenkins-e2e-minion-a6cx

@kubernetes/goog-node @dchen1107 @yujuhong 

Logs from that Kubelet:
https://pantheon.corp.google.com/m/cloudstorage/b/kubernetes-jenkins/o/logs/kubernetes-kubemark-gce/6082/artifacts/130.211.134.119%3A22-kubelet.log

It may be something already known - but it would be great if someone can look into it. @dchen1107 ?

**Comment by yujuhong** [comment]
The cause seems to be that kubelet stops running syncloop. The reconciliation is just a red herring because kubelet is not updating pods anymore. The syncloop hang causes /healthz failure:
```
I0229 18:36:57.348255    3429 server.go:1086] GET /healthz: (91.598Âµs) 500
goroutine 360 [running]:
k8s.io/kubernetes/pkg/httplog.(*respLogger).recordStatus(0xc208347c00, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:214 +0xa6
k8s.io/kubernetes/pkg/httplog.(*respLogger).WriteHeader(0xc208347c00, 0x1f4)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/httplog/log.go:193 +0x32
net/http.Error(0x7feb7da61e88, 0xc208347c00, 0xc208951f20, 0x58, 0x1f4)
	/usr/src/go/src/net/http/server.go:1274 +0x9a
k8s.io/kubernetes/pkg/healthz.funcÂ·002(0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/healthz/healthz.go:109 +0x46e
net/http.HandlerFunc.ServeHTTP(0xc208490330, 0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	/usr/src/go/src/net/http/server.go:1265 +0x41
github.com/emicklei/go-restful.funcÂ·011(0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:278 +0xd3
net/http.HandlerFunc.ServeHTTP(0xc20837d3c0, 0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	/usr/src/go/src/net/http/server.go:1265 +0x41
net/http.(*ServeMux).ServeHTTP(0xc208376f00, 0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	/usr/src/go/src/net/http/server.go:1541 +0x17d
github.com/emicklei/go-restful.Container.ServeHTTP(0x0, 0x0, 0x0, 0xc2084aae80, 0xb, 0x10, 0xc208376f00, 0x0, 0x2e73ec0, 0x0, ...)
	/go/src/k8s.io/kubernetes/Godeps/_workspace/src/github.com/emicklei/go-restful/container.go:264 +0x4f
k8s.io/kubernetes/pkg/kubelet/server.(*filteringContainer).ServeHTTP(0xc20837d340, 0x7feb7da61e88, 0xc208347c00, 0xc2088c7ba0)
	<autogenerated>:30 +0x8a
k8s.io/kubernetes/pkg/kubelet/server.(*Server).ServeHTTP(0xc20823f9c0, 0x7feb7da61e50, 0xc2087fed20, 0xc2088c7ba0)
	/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/server/server.go:1085 +0x13d
net/http.serverHandler.ServeHTTP(0xc2084d1e00, 0x7feb7da61e50, 0xc2087fed20, 0xc2088c7ba0)
	/usr/src/go/src/net/http/server.go:1703 +0x19a
net/http.( [[curl/7.26.0] 127.0.0.1:50769]
```

@Random-Liu, could you take a look, thanks? Let me know if you have too many things on your plate.

**Comment by Random-Liu** [comment]
Sure. I'll look into it.

**Comment by yujuhong** [comment]
@Random-Liu, sorry I misread the issue. Since this is a real kubelet, it's most likely caused by a hung docker request. What we really should do is adding timeout to the docker client in the long run. Don't spend time trying to reproduce this. Let's wait until @dchen1107's PR to properly shutdown/restart docker to see if that helps.

**Comment by Random-Liu** [comment]
Kubelet hangs since:
`I0229 18:31:53.756609    3429 kubelet.go:2421] SyncLoop (housekeeping)`

While, PLEG didn't report anything except several:
`E0229 18:31:47.064312    3429 generic.go:164] GenericPLEG: Unable to retrieve pods: cannot connect to Docker endpoint`
After that it becomes totally silent.

Then kubelet was detected not healthy and was restarted:
`I0229 18:36:57.348255    3429 server.go:1086] GET /healthz: (91.598Âµs) 500`
```
2016-02-29 18:31:48,790 INFO success: docker entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
2016-02-29 18:36:57,354 INFO exited: kubelet (exit status 2; expected)
2016-02-29 18:36:58,356 INFO spawned: 'kubelet' with pid 4303
2016-02-29 18:36:59,399 INFO success: kubelet entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
```

After kubelet was restarted, it started to keep complaining:
`I0229 18:37:39.342791    4318 kubelet.go:2341] skipping pod synchronization - [container runtime is down]`

And the last docker daemon log is:
`time="2016-02-29T18:31:53.952401781Z" level=error msg="HTTP Error" err="No such image: gcr.io/google_containers/kube-proxy:84f5a1be5b6aa26bd5a458fdebecc632" statusCode=404`

In summary, the final conclusion is that docker seems to be unresponsive... :)

**Comment by gmarek** [comment]
It happened on my own 100 node cluster. It happens way to often to ignore this if we want to claim that we support 1000 node ones. cc @davidopp

**Comment by yujuhong** [comment]
@gmarek, do you have any log from the failure (kubelet, docker, kern). That'd definitely help.

We don't see this happening in jenkins enough to gather more information, so we don't have a reliable repro yet.

**Comment by yujuhong** [comment]
It's obvious that docker is in a bad state that kubelet wasn't able to perform any meaningful operations. In fact, cadvisor didn't even finish initialization after kubelet restarts.
However, there is no suspicious messages in kern.log, so I think docker may be fixable through restarts. We need a better docker health in the daemon babysitter for docker.

**Comment by gmarek** [comment]
Both me and @wojtek-t are seeing this quite frequently. I'll grab logs for you tomorrow if it's still needed.

**Comment by yujuhong** [comment]
> Both me and @wojtek-t are seeing this quite frequently. I'll grab logs for you tomorrow if it's still needed.

Sure. Thanks! If we can get docker log with `--log-level=info`, that'd be great.

**Comment by gmarek** [comment]
I'm seeing this in gce soak cluster now. I guess it's better than logs:)

**Comment by yujuhong** [comment]
@gmarek, looking now. Thanks!

**Comment by yujuhong** [comment]
node: jenkins-e2e-minion-ltx6
kubelet version: `kubernetes_build_info{gitCommit="e7d0f06c4aa4375034542b376fbf4c1f16d1fecc",gitTreeState="clean",gitVersion="v1.2.0-alpha.8.709+e7d0f06c4aa437",major="1",minor="2+"} 1`

Basically, the node hit https://github.com/docker/docker/issues/9605 (a known issue), and docker hung. There is not much we can do other than rebooting the node. 

kern.log:
```
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.120110] INFO: task docker:18458 blocked for more than 120 seconds.
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.126889]       Tainted: G         C    3.16.0-4-amd64 #1
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.132580] "echo 0 > /proc/sys/kernel/hung_task_timeout_secs" disables this message.
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.140743] docker          D ffff8800b7509748     0 18458      1 0x00000000
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.152365]  ffff8800b75092f0 0000000000000082 0000000000012f00 ffff88020c69bfd8
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.167350]  0000000000012f00 ffff8800b75092f0 ffffffff818bb4a0 ffff88020c69bd80
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.182358]  ffffffff818bb4a4 ffff8800b75092f0 00000000ffffffff ffffffff818bb4a8
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.197352] Call Trace:
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.201318]  [<ffffffff81510915>] ? schedule_preempt_disabled+0x25/0x70
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.209434]  [<ffffffff815123c3>] ? __mutex_lock_slowpath+0xd3/0x1c0
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.217398]  [<ffffffff815124cb>] ? mutex_lock+0x1b/0x2a
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.224220]  [<ffffffff814175bc>] ? copy_net_ns+0x6c/0x130
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.231260]  [<ffffffff8108bdf4>] ? create_new_namespaces+0xf4/0x180
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.239097]  [<ffffffff8108beec>] ? copy_namespaces+0x6c/0x90
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.246355]  [<ffffffff810654f6>] ? copy_process.part.25+0x966/0x1c30
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.254303]  [<ffffffff81066991>] ? do_fork+0xe1/0x390
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.260955]  [<ffffffff811c442c>] ? __alloc_fd+0x7c/0x120
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.267896]  [<ffffffff81514079>] ? stub_clone+0x69/0x90
Mar  1 23:55:56 jenkins-e2e-minion-ltx6 kernel: [44880.274689]  [<ffffffff81513d0d>] ? system_call_fast_compare_end+0x10/0x15
```

supervisord.log shows kubelet being killed repeatedly, with no sign of docker being killed.
However, if you look at `var/log/supervisor/docker-stdout.log`, you will see the script tried killing docker and waiting for the docker to shutdown (which never happened):
```
Stopping Docker: dockerNo process in pidfile '/var/run/docker.pid' found running; none killed.
 failed!
Starting Docker: docker.
waiting 30s for startup
Docker failed!
Stopping Docker: docker.
waiting clean shutdown
waiting clean shutdown
<skipped>
waiting clean shutdown
waiting clean shutdown
waiting clean shutdown
waiting clean shutdown
```

**Comment by gmarek** [comment]
Are we restarting Kubelet when docker dies, or there's some other reason why supervisor does this? I imagine that the correct behavior in such case is keeping the Kubelet alive and posting some kind of NotReady status with the description of the problem.

**Comment by yujuhong** [comment]
I agree that we should try decoupling kubelet and docker liveness check. As a matter of fact, I removed the docker healthz/ in kubelet for this reason.
Currently, kubelet healthz endpoint checks if syncloop is running, and syncloop wasn't being run because docker was not ready.  A right step forward is to log syncloop timestamp as long as the loop is not hung.

**Comment by yujuhong** [comment]
@vishh, AFAIK, `cadvisor.Start()` will fail if it cannot talk to docker, and it will not retry even if docker comes back. Is that true?
If so, we'll have no option but to kill kubelet even if docker becomes ready, until we change cadvisor's initialization behavior.

**Comment by yujuhong** [comment]
@k8s-oncall, I rebooted jenkins-e2e-minion-ltx6 in the gce-soak suite due to reason stated in https://github.com/kubernetes/kubernetes/issues/22259#issuecomment-191810457

**Comment by gmarek** [comment]
Is it possible to do it for v1.2 (as a cherry-pick, as I don't believe it's possible to do in a day...)? I, for one, think that it's important to handle this is some reasonable fashion (i.e. at least emit a message that will tell user what to do if it happens - and judging by the number of times we run into it, it'll happen a lot).

@davidopp @alex-mohr @roberthbailey @bgrant0607

**Comment by dchen1107** [comment]
AFAIK, there are two separate issues:

* One is syncloop takes long time, but eventually it finishes. We are working on reproducing this issue, and finding the root cause. The root cause could be docker operations are taking a long time, a bug in kubelet, etc. There is no kernel issue with hung task. The node can be self-recovered by either restarting docker or restarting kubelet. Can we use this issue to track that effort? 
* Another one is hitting docker/docker#9605. We used to hit this a lot caused by unregister_netdevice (#20096), and we did mitigate this by workaround the hairpin mode. Once the node run into this issue, the only way to recovery is rebooting the node. I believe some userland behavior triggers this, just like what we were doing with hairpin mode before, but I haven't figured out the use pattern for rest failure yet. Can we use #21866 to track the effort?

**Comment by bgrant0607** [comment]
@gmarek Yes, we expect to cherrypick some PRs after cutting the branch.

**Comment by wojtek-t** [comment]
@yujuhong - what's the status of it?

**Comment by yujuhong** [comment]
I discussed with @dchen1107 for the possibility of decoupling kubelet's and docker's health check, so that we can allow kubelet to report node not ready when docker is down. In order to do so, we will need to modify docker's health check to detect hangs (e.g., using `docker ps`) and put in more robust mechanism (e.g., retries and failure threshold) to prevent killing docker randomly. This may be too risky for v1.2, given that we won't have enough time for testing. @dchen1107, do you want to remove this from the v1.2 milestone?

**Comment by gmarek** [comment]
cc @bgrant0607 @davidopp - as this seriously impacts usability of the system.

**Comment by wojtek-t** [comment]
@gmarek - the question is whether it really does... I didn't see that failure at least since Friday, which means that since we are running scalability suite ~ every hour, that gives us ~120 runs times 100 nodes / run -> 12.000 nodes without such failure. So maybe it's not that serious now...

**Comment by gmarek** [comment]
That's true, but I'd like to understand why it got better, and not assume that it was a very, very bad luck last week.

**Comment by dchen1107** [comment]
Like what I mentioned above and some other issues, I spent more than one days to reproduce this issue, and failed. I did hit some other issues, one of them is  #22480 which is addressed. Also our jenkins builds look good. Changing babysitter for docker daemon to docker ps instead of docker version is too risky for 1.2, especially which might result more unnecessary killing for docker. It also involve kubelet initialization changes. Honestly I am not convinced with the initial triage conclusion of kubelet syncLoop hung. :-) 

I am closing this one for now.

**Comment by yujuhong** [comment]
> Honestly I am not convinced with the initial triage conclusion of kubelet syncLoop hung. :-)

What made you say so? Whether it was caused by docker/kernel or not, kubelet stopped running the sync loop in the initial failure :-)

**Comment by dchen1107** [comment]
Because:

*  Initial comment: https://github.com/kubernetes/kubernetes/issues/22259#issue-137508309 only tells me one of static pods on the node is pending.
* Comment you made at https://github.com/kubernetes/kubernetes/issues/22259#issuecomment-190829856 only tells kubelet failed with /healthz. But the timeline is after test failure. I ruled out this comment.
* Comment @Random-Liu made at https://github.com/kubernetes/kubernetes/issues/22259#issuecomment-190975976 tells me that kube-proxy image is not available. This could be kube-proxy tar file is not copied over or could docker load hang. 

Anyway, none of above comments tell me kubelet syncLoop is really hang. Then rest comments are talking about different issues, for example, docker process is hang forever due to deadlock in kernel.

**Comment by yujuhong** [comment]
> Comment you made at #22259 (comment) only tells kubelet failed with /healthz. But the timeline is after test failure. I ruled out this comment.

kubelet will fail /healthz only if sync loop hangs. I checked the log at the time, and it was clear that kubelet was inactive (not printing any message in sync loop) for a while. The healthz/ failure was reported after the test already failed because kubelet only fails the healthz/ if sync loop has not been run in 5 minutes. Not that this is important now, but I think it's a hang due to docker/kernel issues.
